### PR TITLE
Add API to compile from source string

### PR DESCRIPTION
**Pull request by alan-baker**
New API analogous to clBuildProgram that compiles from source strings.

Contributes to #242

* Refactored ArgKind to expose the enum externally
 * Formalized PodUBO as an ArgKind, but plumbing is minimal still
* Added an initial version of data structure for descriptor map entry
 * ostream operator writes current desriptor map
* new api populates SPIR-V and descriptor map in memory
 * SPIRVProducerPass reworked to populate the descriptor map data structure

Tested against unit tests and clvk (see https://github.com/alan-baker/clvk/tree/use-clspv-api)

### CODE REVIEW

**Comment by ericberdahl** [comment]
I'm becoming confused about the purpose and structure of this project. clspv self-identifies as a prototype compiler for a subset of OpenCL C to Vulkan compute shaders. That's very different than providing a library or API for applications to use to compile and build OpenCL C programs as Vulkan compute shaders. I'm concerned that the purpose, goal, and direction of clspv is being subtly changed in ways that will eventually prove to be conflicting.

The goal of providing APIs and libraries that allow applications to more easily use and consume the output artifacts of clspv's compilation is a good goal. The goal of providing APIs and libraries that allow applications to more easily use clspv in a manner similar to the way they use existing OpenCL APIs is a good goal. Neither of those goals are tightly aligned with the goal of creating a tool that compiles a subset of OpenCL C to Vulkan compute shaders.

**Comment by kpet** [comment]
I've pulled this locally and tried to run the OpenCL conformance tests. Out of give or take 400 tests usually passing for me in my quick set, only 19 pass with this change and your clvk branch. Most of the test binaries just crash before reporting results. The first two I've looked at crash with the same message:

```
clspv: for the -simplifycfg-sink-common option: may only occur zero or one times!
```

This confirms my fear: crashing or terminating the process will no longer be acceptable. A quick grep shows that there are at least 75 calls to `llvm_unreachable` that will need to be taken care of (and a few more asserts). Asking for these to be turned into controlled errors in one go just isn't realistic so I suggest we keep the compilation API as a build option in clvk and work through these over time. This leads me to the next question: have you given any thought as to how you want to handle errors in passes? I've tried to play with some of the error handling features of llvm (`install_fatal_error_handler` + `report_fatal_error`) while trying to solve #186 but this becomes a bit more involved when everything is running in the same process.

**Comment by alan-baker** [comment]
Ok, I've been testing with the simple tests so far. I will look into invoking a compile multiple time in a process and track down those errors.

File: include/clspv/Compiler.h [hunk 1]
```diff
@@ -12,7 +12,37 @@
...
+#ifndef CLSPV_INCLUDE_CLSPV_COMPILER_H_
+#define CLSPV_INCLUDE_CLSPV_COMPILER_H_
+
+#include <string>
+#include <vector>
+
+#include "clspv/DescriptorMap.h"
+
 namespace clspv {
+struct Result {
```
Comment (kpet) [inline include/clspv/Compiler.h]: Have you considered using `StringError` and friends?

File: include/clspv/DescriptorMap.h [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    uint32_t local_element_size;
+
+    // POD data.
+    uint32_t pod_offset;
+    uint32_t pod_arg_size;
+  } kernel_arg_data;
+
+  struct ConstantData {
+    ArgKind constant_kind;
+    std::vector<uint8_t> hex_bytes;
```
Comment (kpet) [inline include/clspv/DescriptorMap.h]: Probably worth renaming `hex_bytes`.

File: lib/Compiler.cpp [hunk 3]
```diff
@@ -151,176 +157,192 @@ static llvm::cl::opt<bool> verify("verify", llvm::cl::init(false),
 
 // Populates |SamplerMapEntries| with data from the input sampler map. Returns 0
 // if successful.
-int ParseSamplerMap(llvm::SmallVectorImpl<std::pair<unsigned, std::string>> *SamplerMapEntries) {
-  if (!SamplerMap.empty()) {
+int ParseSamplerMap(
+    const std::string &sampler_map,
+    llvm::SmallVectorImpl<std::pair<unsigned, std::string>> *SamplerMapEntries,
+    llvm::raw_ostream *error_str) {
```
Comment (kpet) [inline lib/Compiler.cpp]: Passing by it by reference would make it easier to use.

Comment (alan-baker) [reply lib/Compiler.cpp]: [Google style guide](https://google.github.io/styleguide/cppguide.html) says all references must be const-qualified.

**Comment by kpet** [comment]
@alan-baker That's made a big difference already. 140 tests passing for me. The binaries now seem to be crashing mostly in the passes.

File: include/clspv/Compiler.h (hunk 1 above)
Comment (alan-baker) [reply include/clspv/Compiler.h]: Right now the API doesn't expose LLVM headers and I'd prefer to keep it that way.

File: include/clspv/DescriptorMap.h (hunk 2 above)
Comment (alan-baker) [reply include/clspv/DescriptorMap.h]: Good suggestion. I've renamed it to initializer.

**Pull request merged**